ten-simple-rules-dockerfiles.Rmd

```diff
@@ -562,70 +562,64 @@ RUN pip install \
 
 \normalsize
 
-serves as more clear documentation in a `Dockerfile` than a `requirements.txt` file that lists the same:
+Alternatively, versions may be specified in a separate dependency file (e.g., `requirements.txt` or `environment.yml`) and `COPY`ied to the image for installation:
 
 \footnotesize
 
 ```bash
+COPY requirements.txt .
 RUN pip install -r requirements.txt
 ```
 
 \normalsize
 
-This modularisation may pose a risk for understandability and consistency, but it can be mitigated by carefully managing all these files in the same version-controlled project.
+This modularization may reduce readability, but provides more flexibility in facilitating different ways of building a reproducible environment, provided the dependency file is under version control in the same repository (see \ruleref{rule:publish}).
 You can also use package managers to install software from source code `COPY`ied into the image (see&nbsp;\ruleref{rule:mount}).
-And, finally, you can use many package managers to install software obtained from code management repositories, e.g., installing a specific tool identified by a GitHub version tag or commit hash.
-Be aware of the risk of depending on such repositories, especially if they are out of your control, but the installation command with a full URL should allow readers of your `Dockerfile` to dig deeper if problems arise.
-Such an installation from a source gives you much freedom, including the freedom to shoot yourself in the foot!
+Finally, you can use many package managers to install software from source obtained from external code management repositories, e.g., installing a tool from a specific GitHub branch or version tag.
```

Comment:
I'd prefer to leave this at "version tag or commit hash", because installing a branch means installing a moving target, not pinning a specific version.

Reply:
Done!